WordPress/src/main/java/org/wordpress/android/ui/reader/discover/interests/ReaderInterestsViewModel.kt

```diff
@@ -1,57 +1,151 @@
 import androidx.lifecycle.LiveData
 import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
 import kotlinx.coroutines.launch
+import org.wordpress.android.R
+import org.wordpress.android.models.ReaderTag
 import org.wordpress.android.models.ReaderTagList
+import org.wordpress.android.ui.reader.discover.interests.ReaderInterestsViewModel.DoneButtonUiState.DoneButtonDisabledUiState
+import org.wordpress.android.ui.reader.discover.interests.ReaderInterestsViewModel.DoneButtonUiState.DoneButtonEnabledUiState
+import org.wordpress.android.ui.reader.discover.interests.ReaderInterestsViewModel.DoneButtonUiState.DoneButtonHiddenUiState
 import org.wordpress.android.ui.reader.repository.ReaderTagRepository
+import org.wordpress.android.viewmodel.Event
 import javax.inject.Inject
 
 class ReaderInterestsViewModel @Inject constructor(
     private val readerTagRepository: ReaderTagRepository
 ) : ViewModel() {
     var initialized: Boolean = false
 
-    private val _uiState = MutableLiveData<UiState>()
+    private val _uiState: MutableLiveData<UiState> = MutableLiveData(
+        UiState(listOf(), ReaderTagList(), DoneButtonHiddenUiState)
+    )
     val uiState: LiveData<UiState> = _uiState
 
+    private val _navigateToDiscover = MutableLiveData<Event<Unit>>()
+    val navigateToDiscover: LiveData<Event<Unit>> = _navigateToDiscover
+
+    private val selectedInterests = ReaderTagList()
+
     fun start() {
-        if (initialized) return
+        if (initialized) {
+            val uiState = uiState.value as UiState
+            updateUiState(uiState.copy(interestsUiState = getInterestsUiStateWithSelectedStates()))
+            return
+        }
         loadInterests()
     }
 
     private fun loadInterests() {
         viewModelScope.launch {
             val tagList = readerTagRepository.getInterests()
             if (tagList.isNotEmpty()) {
-                updateUiState(UiState(transformToInterestsUiState(tagList), tagList))
+                updateUiState(
+                    UiState(
+                        transformToInterestsUiState(tagList),
+                        tagList,
+                        getDoneButtonUiState()
+                    )
+                )
                 if (!initialized) {
                     initialized = true
                 }
             }
         }
     }
 
+    fun onInterestAtIndexToggled(index: Int, isChecked: Boolean) {
+        uiState.value?.let {
+            val enableDoneButton = selectedInterests.isEmpty() && isChecked
+            val disableDoneButton = selectedInterests.size == 1 && !isChecked
+
+            updateSelectedInterests(it.interests[index])
+
+            if (enableDoneButton || disableDoneButton) {
+                val uiState = uiState.value as UiState
+                updateUiState(
+                    uiState.copy(
+                        interestsUiState = getInterestsUiStateWithSelectedStates(),
```

Comment:
AFAIU we are updating the uiState only when the state of the done buttons changes. It means that there are scenarios when the uiState doesn't correspond to the state of the UI (selected items). It feels like it goes against one of the main benefits of using UiState. Is this intentional? Can't we always keep it up-to-date? Am I missing something? Wdyt?

Reply:
Refactored here: bfaaee9